engine/collection/ingest/engine.go

```diff
@@ -65,97 +72,147 @@ func New(
+		pendingTransactions:  pendingTransactions,
+		messageHandler:       handler,
 		pools:                pools,
 		config:               config,
 		transactionValidator: transactionValidator,
 	}
 
+	e.ComponentManager = component.NewComponentManagerBuilder().
+		AddWorker(e.processQueuedTransactions).
+		Build()
+
 	conduit, err := net.Register(engine.PushTransactions, e)
 	if err != nil {
 		return nil, fmt.Errorf("could not register engine: %w", err)
 	}
-
 	e.conduit = conduit
 
 	return e, nil
 }
 
-// Ready returns a ready channel that is closed once the engine has fully
-// started.
-func (e *Engine) Ready() <-chan struct{} {
-	return e.unit.Ready()
-}
+// Process processes a transaction message from the network and enqueues the
+// message. Validation and ingestion is performed in the processQueuedTransactions
+// worker.
+func (e *Engine) Process(channel network.Channel, originID flow.Identifier, event interface{}) error {
+	select {
+	case <-e.ComponentManager.ShutdownSignal():
+		return component.ErrComponentShutdown
+	default:
+	}
 
-// Done returns a done channel that is closed once the engine has fully stopped.
-func (e *Engine) Done() <-chan struct{} {
-	return e.unit.Done()
+	err := e.messageHandler.Process(originID, event)
+	if err != nil {
+		if engine.IsIncompatibleInputTypeError(err) {
+			e.log.Warn().Msgf("%v delivered unsupported message %T through %v", originID, event, channel)
+			return nil
+		}
+		return fmt.Errorf("unexpected error while processing engine message: %w", err)
+	}
+	return nil
 }
 
-// SubmitLocal submits an event originating on the local node.
-func (e *Engine) SubmitLocal(event interface{}) {
-	e.unit.Launch(func() {
-		err := e.process(e.me.NodeID(), event)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
+// ProcessTransaction processes a transaction message submitted from another
+// local component. The transaction is validated and ingested synchronously.
+// This is used by the GRPC API, for transactions from Access nodes.
+func (e *Engine) ProcessTransaction(tx *flow.TransactionBody) error {
+	return e.onTransaction(e.me.NodeID(), tx)
 }
 
-// Submit submits the given event from the node with the given origin ID
-// for processing in a non-blocking manner. It returns instantly and logs
-// a potential processing error internally when done.
-func (e *Engine) Submit(channel network.Channel, originID flow.Identifier, event interface{}) {
-	e.unit.Launch(func() {
-		err := e.process(originID, event)
-		if err != nil {
-			engine.LogError(e.log, err)
+// processQueuedTransactions is the main message processing loop for transaction messages.
+func (e *Engine) processQueuedTransactions(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-e.messageHandler.GetNotifier():
+			err := e.processAvailableMessages(ctx)
+			if err != nil {
+				// if an error reaches this point, it is unexpected
+				ctx.Throw(err)
```

Comment:
Generally speaking `Throw` should terminate current goroutine but I would still add a `return` after to be safe